frontend/dispatcher.lua

```diff
@@ -254,6 +244,40 @@ function Dispatcher:addSubMenu(menu, location, settings)
                 })
             end
         end
+    end
+end
+
+--[[--
+Add a submenu to edit which items are dispatched
+arguments are:
+    1) self
+    2) the table representing the submenu (can be empty)
+    3) the name of the parent of the settings table (must be a child of self)
+    4) the name of the settings table
+example usage:
+    Dispatcher.addSubMenu(self, sub_items, "profiles", "profile1")
+--]]--
+function Dispatcher:addSubMenu(menu, location, settings)
+    if not Dispatcher.initialized then Dispatcher:init() end
+    table.insert(menu, {
+        text = _("None"),
+        separator = true,
+        checked_func = function()
+            return next(self[location][settings]) == nil
+        end,
+        callback = function(touchmenu_instance)
+            self[location][settings] = {}
+            if touchmenu_instance then touchmenu_instance:updateItems() end
+        end,
+    })
+    for _,section in ipairs({"Device", "Crengine",}) do
+        local submenu = {}
+         Dispatcher.addItem(self, submenu, location, settings, section)
```

Comment:
nit

Reply:
(I guess the nit was about the indentation.)
But there's also `Dispatcher.addItem(self, ...)` which looks like it could be `Dispatcher:addItem(...)` (with a colon and without self, as usually when calling a method) - unless there's something else involved :)

Reply:
I added a comment:

Must be executed in the context of the caller to add items to the callers menu.
therefore declared as Dispatcher:addSubMenu(menu, location, settings)
but should be called as Dispatcher.addSubMenu(self, menu, location, settings)

the same applies to addItem

Reply:
I'm not sure I get it, and what you mean with "context".
What object exactly is `self` when these are called in real-life?
A Dispatcher instance ? The Dispatcher static class object ? Something not a Dispatcher ? :)
If `self` is something else than a Dispatcher thingy, better to declare and use them with a dot so there's no confusion, and name the first argument something other than `self`.

Reply:
self is whoever is calling Dispatcher, right now that is only Profiles but in the future also ReaderGesture.

see:
https://github.com/koreader/koreader/blob/509ee7bb8612ddb370f4f61d3e4a3209d7ec46ad/plugins/profiles.koplugin/main.lua#L107

i changed it to be declared as `Dispatcher.addSubMenu(caller, menu, location, settings)`